vortex-compute/src/filter/buffer_mut.rs

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright the Vortex contributors
+
+use core::ptr;
+
+use vortex_buffer::BufferMut;
+use vortex_mask::Mask;
+
+use crate::filter::Filter;
+
+impl<T: Copy> Filter for &mut BufferMut<T> {
+    type Output = ();
+
+    fn filter(self, selection_mask: &Mask) {
+        assert_eq!(
+            selection_mask.len(),
+            self.len(),
+            "Selection mask length must equal the buffer length"
+        );
+
+        match selection_mask {
+            Mask::AllTrue(_) => {}
+            Mask::AllFalse(_) => self.clear(),
+            Mask::Values(values) => {
+                // We choose to _always_ use slices here because iterating over indices will have
+                // strictly more loop iterations than slices, and the overhead over batched
+                // `ptr::copy(len)` is not worth it.
+                let slices = values.slices();
+
+                // SAFETY: We checked above that the selection mask has the same length as the
+                // buffer.
+                let new_len = unsafe { filter_slices_in_place(self, slices) };
+
+                // Truncate the buffer to the new length.
+                self.truncate(new_len);
```

Comment:
Why not also `unsafe { self.set_len(new_len) }`

Reply:
In fact, you really need to do that first.

If you add `vortex-compute` to miri CI check I believe you'll get a failure: https://github.com/vortex-data/vortex/blob/86f3bf9caedf71f99810786ee677bdafa48975fb/.github/workflows/ci.yml#L620